### ISSUE TITLE
CI4: Upgrade to CodeIgniter 4.4.5

### ISSUE THREAD

**Issue by daN4cat** [opensourcepos/opensourcepos#842]
Code Igniter 4 is entering Alpha release: https://github.com/bcit-ci/CodeIgniter4

We should monitor progress and upgrade when an official ans stable release is pushed out.

**Comment by objecttothis** [comment]
@daN4cat @jekkos I wouldn't mind tackling this.  I was just reading the documentation (https://codeigniter.com/user_guide/installation/upgrade_4xx.html) and while it's more work than other incremental changes, it appears to be primarily syntax changes.  Thoughts?

**Comment by objecttothis** [comment]
Implications would mean that our minimum PHP version would be 7.3.x.  Currently OSPOS has a 7.2.x minimum for PHP.  According to CI migrations after 3.x to 4.x are simpler.

**Comment by objecttothis** [comment]
@opensourcepos/org an optional CI4 change in the Views is to change the `echo` syntax from

`<?php echo $title; ?>`
to
`<?= $title; ?>`

So for example, `<?php echo current_language_code(); ?>` would become `<?= current_language_code(); ?>`

IMO this is cleaner, therefore easier to read, but it is optional, so if we veto, then that's OK too.  Alternately, I could just do the minimum to port over to CI4 and we treat the optional changes as a question to be answered separately.

Thoughts?

**Comment by jekkos** [comment]
I've seen this pattern before so here and there, don't think it's new in CI4? Let's do the minimum first and then these kind of 'syntax' cleanups can be done after?

**Comment by odiea** [comment]
Let me know if there is anything I can do to help out

**Comment by objecttothis** [comment]
> 
> 
> Let me know if there is anything I can do to help out

Once I get it to a place where it builds I'll need someone to help me break it ;)

**Comment by odiea** [comment]
Is this all going into 1 request or application?

**Comment by odiea** [comment]
Also are you upgrading  Bootstrap

**Comment by daN4cat** [comment]
@jekkos unless released I would kick out 3.4.0 or whatever is the version and make CI4 for a new release. 
It's either OSPOS 4.0.0 or 3.5.0.

I guess a version 4 would mean more refactoring like BS4, adding REST APIs and maybe a Single Page Application based in Angular or React.

Not sure what we can really achieve here.

**Comment by objecttothis** [comment]
> Also are you upgrading Bootstrap

If compatible I will not upgrade bootstrap here but leave that for someone to do in a later pr.

**Comment by objecttothis** [comment]
> @jekkos unless released I would kick out 3.4.0 or whatever is the version and make CI4 for a new release.
> It's either OSPOS 4.0.0 or 3.5.0.
> 
> I guess a version 4 would mean more refactoring like BS4, adding REST APIs, OAuth and maybe a Single Page Application based in Angular or React.
> 
> Not sure what we can really achieve here.

I was anticipating that this would be a release after 3.4.0 and not hold that up.

My personal preference would be to make it a 3.5.0 release because even though it's a massive refactoring, and will provide many ease of life improvements it isn't fundamentally adding major functions to the software.  It does set the stage for several major changes.

I am interested in helping with Restful API Interfaces. It something I need, so I'm motivated.

**Comment by objecttothis** [comment]
> Is this all going into 1 request or application? I suppose I will need to build it with composer

As for building, I'm going to get it to the point where it builds, but there may be some tweaks made to the CI/CD setup on the repo so that it builds and deploys here. Hopefully we can get it to a place where you won't need to build locally.

**Comment by odiea** [comment]
I just wanted to make sure that the versions that I use to compile are the same as what the standard is. Apples to Apples as the saying goes.

**Comment by jekkos** [comment]
@daN4cat I was also reflecting on this. What should be the scope of 3.4? Typically indeed we have at least one new bigger feature. I was maybe thinking of this person attributes PR as a last addition.

A lot of tickets have 3.4 milestone set, but this isn't realistic at all.

**Comment by jekkos** [comment]
Regarding the REST APIs, I think we already have REST endpoints for CRUD.

(create = save, read = search, update = save, delete = delete)

All of these methods return JSON so IF one wants to build a new frontend, then we can just use those, no?

**Comment by objecttothis** [comment]
The thought dawned on me that this will go faster if more than one person is working on this.  One of the changes that will take a good amount of just working through each query is the syntax for query builder has changed (simplified in some ways) between CI3 and CI4.  Does anyone want to work through some or all of the queries and translate them from one to the other?  https://codeigniter.com/user_guide/database/query_builder.html  

No worries if not.  I will keep plugging away at it.  For the most part, none of this so far has been a big quandary, but just a matter of going through it all.

**Comment by objecttothis** [comment]
This might be a conversation for another day, but it appears that CodeIgniter has moved away from underscored_variable_names to camelCase.  I think the change is meant to bring CI inline with PSR coding standards.  That said, I understand that the conversation is widely argued and I don't really feel that strongly about changing for the sake of changing, but our codebase will now have camelCase variables and underscore_variables... which is more annoying to me.

**Comment by objecttothis** [comment]
@daN4cat @jekkos If the upgrade requires any changes to the CI/CD scripts, then I'll likely need your help with that part.  There are several folder name changes in the project, so I imagine some amount of tweaking will be needed.

**Comment by jekkos** [comment]
Yes I think there are some commands that refer to the application/config path for example, is this different in CI4? I can try to update the .travis.yml file in your branch to fix this.

Also the translations are downloaded and installed so that might need to change as well.

**Comment by objecttothis** [comment]
@jekkos good point about translations.  Yeah the config file that governs where to download the files to has to change.  Instead of application/language the base folder is now app/Language.  I don't think the files themselves will need to change, but I haven't gotten that far yet.

**Comment by objecttothis** [comment]
I take that back.  The formatting of the language files is different.  https://codeigniter4.github.io/userguide/outgoing/localization.html  That could be a conversion that someone can take if they want.  If not, that's OK too.  It's mostly a bunch of find/replace... That will however affect weblate... I'm not sure if it uses templates to build the files when it downloads them, but those templates will need to change.

Here is what the new language file format looks like.
```
return [
    'errorEmailMissing'    => 'You must submit an email address',
    'errorURLMissing'      => 'You must submit a URL',
    'errorUsernameMissing' => 'You must submit a username',
    'nested'               => [
        'error' => [
            'message' => 'A specific error message',
        ],
    ],
];
```

The usages will also need to be refactored: `echo lang('Errors.errorEmailMissing');` is cleaner at least :)

**Comment by objecttothis** [comment]
At least the changes to the way localization works makes it easier to call translations from various different files when occasionally needed.

**Comment by jekkos** [comment]
I see, the new format is just a little different in this case, it just creates a new array instead of appending translations to an existing one, should be feasible to change that.

**Comment by jekkos** [comment]
I suppose we need to fetch translations from this repo now

https://github.com/codeigniter4/translations/tree/develop/Language

PHP version 7.3 or newer is required, with the *intl* extension and *mbstring* extension installed.
It seems that CI build fails now as the composer image does not have intl installed

```
    - codeigniter4/framework v4.1.3 requires ext-intl * -> the requested PHP extension intl is missing from your system.
```

**Comment by jekkos** [comment]
It seems we have also consumed all of the build credits on Travis

```
        Builds have been temporarily disabled for public repositories due to a negative credit balance. Please go to the Plan page to replenish your credit balance or alter your Consume paid credits for OSS setting.

```

I might want to try and see if there is a possibility to build PRs using opt-in instead, as currently it might not make sense to have a build for each commit if you have work ongoing?

**Comment by objecttothis** [comment]
> 
> 
> It seems we have also consumed all of the build credits on Travis
> 
> ```
>         Builds have been temporarily disabled for public repositories due to a negative credit balance. Please go to the Plan page to replenish your credit balance or alter your Consume paid credits for OSS setting.
> ```
> 
> I might want to try and see if there is a possibility to build PRs using opt-in instead, as currently it might not make sense to have a build for each commit if you have work ongoing?

Opt-in or on-demand.  Definitely don't need a build for each commit.

**Comment by objecttothis** [comment]
> 
> 
> I suppose we need to fetch translations from this repo now
> 
> https://github.com/codeigniter4/translations/tree/develop/Language
> 
> PHP version 7.3 or newer is required, with the _intl_ extension and _mbstring_ extension installed.
> It seems that CI build fails now as the composer image does not have intl installed
> 
> ```
>     - codeigniter4/framework v4.1.3 requires ext-intl * -> the requested PHP extension intl is missing from your system.
> ```

That's surprising to me.  I thought intl was already a project requirement for OSPOS.

**Comment by jekkos** [comment]
Yes it is, but now it's probably also a composer dependency. (which might not have been the case before)

**Comment by objecttothis** [comment]
@opensourcepos/org If anyone does take part of this project, let's work in the same branch and just communicate here or in the Pull Request Conversation about what you are taking.  I have the whole workday on Tuesdays and Thursdays to work on this, but like I said before, a project this size will go faster if multiple people are chipping away at it.  I'd be glad to help anyone if they get stuck as I've been pouring over the documentation regarding the changes.  Just make sure you communicate so that no work ends up getting duplicated.

**Comment by objecttothis** [comment]
@jekkos There are a few changes with the language file names and structures that will affect the weblate setup.
- %1 gets replaced with {0}
- All phrases are escaped with ' even if they have {0} inside them.
- ' inside a phrase must be escaped with \
- File name is capitalized and _lang is removed (Attributes.php instead of attributes_lang)
- The file itself just returns an array ( `return [ ... ];`) with no `<php?` at the beginning.
- The index for the phrase has the prefix removed (attributes_confirm_delete becomes confirm_delete).  This is done because the usage will become echo `lang('Attributes.confirm_delete');` and Attributes.attributes_confirm_delete is redundant.
- I've refactored en-US as an example.  I'm not sure the quickest way to refactor all the other ones.

**Comment by objecttothis** [comment]
I also just pushed a commit refactoring the usages of the phrases.

**Comment by objecttothis** [comment]
I finished language conversion and I'm mid-swing with converting the querybuilder calls in the models.  

I haven't touched converting the views yet if anyone wants to take that.  I think it's primarily just $this->load->view(x); becomes echo view(x);.

https://codeigniter.com/user_guide/installation/upgrade_views.html

Its the end of the day for me here, so I'll pick this up again on Thursday.

**Comment by odiea** [comment]
@objecttothis Thanks for all the work that you are doing.

**Comment by objecttothis** [comment]
@jekkos @daN4cat Decision time:
CI4 is namespaced and among the changes, models now extend CodeIgniter\Model which itself extends BaseModel.  This is great, except several of our models contain a `public function save()` and since BaseModel contains a definition for save() it thinks that this method is an override of the BaseModel method.  It complains because the signatures are different.

I think the easiest thing to do would be to refactor the name of these methods.  It wouldn't be difficult to do.

Same thing with delete().  If we do want to rename them, what do we change them to?  I haven't looked at the two functions to see exactly what they are doing.  If they are completing the same behavior, then the proper thing to do would be to rework the functions so that they are actually overriding the baseModel class' method.

**Comment by objecttothis** [comment]
Sorry, it appears that most of these can be resolved by adding the return type to the function definition.

**Comment by jekkos** [comment]
@objecttothis I guess the best fix would be to refactor it then indeed. Do you have any idea what those methods do?

Using phpstorm the rename should be straightforward hopefully.

**Comment by objecttothis** [comment]
We need to sort out exactly how to reference model functions in another model.  I posted a question on stack overflow about this (https://stackoverflow.com/questions/69030352/what-is-the-proper-way-to-access-a-model-function-from-another-model-in-ci4) and I might have gotten it worked out, but I'm not 100% sure until I start testing the application.

Here is what I've done, and at least it's stopped complaining about the model...

```
<?php

namespace App\Models;

use CodeIgniter\Model;
use app\Libraries\Sale_lib;

/**
 * Sale class
 * @property mixed attribute
 * @property mixed config
 * @property mixed customer
 * @property mixed dinner_table
 * @property mixed inventory
 * @property mixed item
 *
 * @property Sale_lib sale_lib
 */
class Sale extends Model
{
	public function __construct()
	{
		$this->attribute = model('Attribute');
		$this->config = model('Config');
		$this->customer = model('Customer');
		$this->dinner_table = model('Dinner_table');
		$this->inventory = model('Inventory');
		$this->item = model('Item');

		$this->sale_lib = new Sale_lib();
	}

```
It's not clear that this is the right way to do it though.  What's confusing to me is where app\Config\Autoload.php is supposed to come into it.  I would have hoped that this file would have prevented all the needs for @property declarations and references in the constructor.

**Comment by jekkos** [comment]
I wonder do we really have so many models depending on each other? It does not sound like good design to mix up different bounded context. 

At least for basic crud operations it should not be the case.

I do agree that controllers might have dependencies on different models, but maybe the models themselves should have near 0 deps

**Comment by objecttothis** [comment]
> 
> 
> I wonder do we really have so many models depending on each other? It does not sound like good design to mix up different bounded context.
> 
> At least for basic crud operations it should not be the case.
> 
> I do agree that controllers might have dependencies on different models, but maybe the models themselves should have near 0 deps

 I agree, however those changes should be outside the scope of this PR or we risk breaking things in more than a few ways.

**Comment by objecttothis** [comment]
@jekkos, I suspect, by nature of what it is, the Sale.php model is likely the worst offender.  Attribute.php, for example, has just one model dependency (Appconfig).

**Comment by jekkos** [comment]
Ok interesting. For the latter I think that we might be able to replace that code with functions from the config helper? Not sure why we need the dependencies in the sale model, it's worth a look.

**Comment by objecttothis** [comment]
> 
> 
> Ok interesting. For the latter I think that we might be able to replace that code with functions from the config helper? Not sure why we need the dependencies in the sale model, it's worth a look.

I think it's because sale touches so much in the database.  Take this line from the (way-too-large) save function:
`$this->giftcard->update_giftcard_value( $splitpayment[1], $cur_giftcard_value - $payment['payment_amount'] );`

When a customer uses a giftcard as part of the payment, it's updating the value of the giftcard so that the customer can't double spend.  Maybe I'm overlooking something, but short of copying and pasting that code into this model, I don't know how we would remove the dependency on the giftcard model.  We could perhaps create a helper and function there that has all the dependencies, but that seems like it's just obfuscating the dependency.

**Comment by jekkos** [comment]
I guess the proper design solution might be to move this logic up to the controller so that we only have atomic crud operations in the model for every entity?

This is really business logic. Indeed if we introduce business logic in the models, we need those dependencies. But that might not be according to the CI design philosophy?

**Comment by jekkos** [comment]
If they do not provide a standard way to introduce dependencies in the models then I guess there is a reason for this

**Comment by objecttothis** [comment]
> 
> 
> I guess the proper design solution might be to move this logic up to the controller so that we only have atomic crud operations in the model for every entity?
> 
> This is really business logic. Indeed if we introduce business logic in the models, we need those dependencies. But that might not be according to the CI design philosophy?

Sounds correct to me.  Software design is still something I'm learning, but moving this business logic into the Controller in another PR would produce the separation you are proposing.

**Comment by objecttothis** [comment]
@jekkos @daN4cat I have been peppering the code with //TODO: when I find something that needs to be fixed but is outside the scope of the CI4 migration.  It will slow things down, but if you prefer, I can create github issues instead.  My tentative plan was to go back through the code after CI4 is stable and 3rd party integrations are working and create separate PR's for each TODO.

**Comment by jekkos** [comment]
Sounds good to me, i think it should be easy to get an overview of the TODOs in the branch this way

**Comment by bhojkamal** [comment]
I pulled branch CodeIgniter_4_upgrade to my local to test, I could not run it.  I got this message 

`Your system folder path does not appear to be set correctly. Please open the following file and correct this: index.php`

**Comment by objecttothis** [comment]
> I pulled branch CodeIgniter_4_upgrade to my local to test, I could not run it. I got this message
> 
> `Your system folder path does not appear to be set correctly. Please open the following file and correct this: index.php`

That's because I'm in the middle of code changes. It won't run right now.

**Comment by bhojkamal** [comment]
@objecttothis, Yes, I see, still need to work more to upgrade CI4. e.g. routes, index.php, App.php, config, .env setting.

**Comment by bhojkamal** [comment]
When we upgrade to codeIgniter 4.1.3, we have to upgrade the bootstrap as well. bootstrap 3 is very old. It is better to upgrade to vanilla javascript as jquery is also removed in bootstrap 5.

**Comment by objecttothis** [comment]
I am now finished with my initial changes and I'm going back through all of the TODO statements that I made to address the uglier parts of the upgrade from CI 3.x to 4.x.  Once that is done I will move on to debugging and making sure that everything is rendering properly.

That said, @bhojkamal I hear you that Bootstrap 3 is very old and agree that it needs to be upgraded.  The problem is that the CI 4.1.3 upgrade is massive and we need to keep feature set changes as separate as possible.  If possible, let's get CI 4.x in a dev build and working to everyone's satisfaction, then merge it into the master.  Once that is done I think we should work on Bootstrap upgrades which when finished will make up the body of an OpenSourcePOS 4.0 release.  @jekkos I think a new bootstrap in addition to CI 4.1.3 would satisfy your desire to make it a 4.0 rather than 3.x, correct?

**Comment by objecttothis** [comment]
@jekkos @daN4cat @opensourcepos/org now that I am done doing the basic changes, I need to focus on these more complex changes that will require some design decisions.  I don't want to make these decisions by myself, but I also don't want the work to stop because I can't get answers to these questions.  Do any of you have time to put into making these design decisions so that I can proceed?  For example, one of them is that the current implementation has the Events\Load_config class using the Appconfig model to pull down all the configuration items from the database and then load them into the CI3 config object.  Then throughout the code, this data is called upon.  I think the primary purpose of this is that it acts as a cache to reduce the number of database calls to get configuration items.

The problem is that the config object in CI4 is now broken up into a couple dozen different config files and corresponding objects.  We need to decide if we continue caching the appconfig (probably a good idea), but if so, do we dump the values into one of the existing objects (config("App") maybe?) or do we create a custom cache object that we then reference in the code instead of database calls every time we need a config setting.

**Comment by jekkos** [comment]
> When we upgrade to codeIgniter 4.1.3, we have to upgrade the bootstrap as well. bootstrap 3 is very old. It is better to upgrade to vanilla javascript as jquery is also removed in bootstrap 5.

Yes the frontend tech stack is quite outdated.. Ideally it is rewritten in angular or react. I think using those toolkits only half of the code would remain. But we might need funding to hire someone this. Like perhaps [a student or something](https://summerofcode.withgoogle.com/get-started)

**Comment by jekkos** [comment]
@objecttothis indeed I think the Bootstrap upgrade will justify a 4.0 really. On the other hand if you look at  the amount of time spent to do the upgrade, the CI bump is quite a significant jump. But the end user won't really notice (aside from some potential bugs or regressions ;)

I think that it's a good idea to try to keep the config values in memory, as they are used very often and running a query for every one of them will probably not be very performant. I think this was indeed something custom done on top of CI3, so now this infrastructure has changed which means that you'll need to rebuild it in some way? does it seem like a lot of work or rather doable?

**Comment by objecttothis** [comment]
> @objecttothis indeed I think the Bootstrap upgrade will justify a 4.0 really. On the other hand if you look at the amount of time spent to do the upgrade, the CI bump is quite a significant jump. But the end user won't really notice (aside from some potential bugs or regressions ;)

Yeah, I would actually be surprised if there aren't a number of bugs and regressions to work through.  We will probably need to have this CI4 as a dev build for some time before releasing a production build. 

> I think that it's a good idea to try to keep the config values in memory, as they are used very often and running a query for every one of them will probably not be very performant. I think this was indeed something custom done on top of CI3, so now this infrastructure has changed which means that you'll need to rebuild it in some way? does it seem like a lot of work or rather doable?

Each of the configs are now stored in separate files with a single .env allowing the quick swapping of connection variables, etc.  CI4 does have the ability to create custom configuration files which is probably the route that we want to go since each config file is mapped automatically to an object and we can then port our current config_loader to that system and keep loading the config into memory.

**Comment by jekkos** [comment]
Ok ideally we leverage the CI4 approach, it might be a matter of running another find/replace to fix this to load everything from one map, or is it more involved?

**Comment by objecttothis** [comment]
> > When we upgrade to codeIgniter 4.1.3, we have to upgrade the bootstrap as well. bootstrap 3 is very old. It is better to upgrade to vanilla javascript as jquery is also removed in bootstrap 5.
> 
> Yes the frontend tech stack is outdated.. Ideally it is rewritten in angular or react. I think using those toolkits only half of the code would remain. But we might need funding to hire someone to do this. Like perhaps [a student or something](https://summerofcode.withgoogle.com/get-started)

https://www.youtube.com/watch?v=Uo3cL4nrGOk

**Comment by objecttothis** [comment]
@jekkos @daN4cat Here's another design decision issue.  In CI4 all models need to extend Model.  In this case Model has a get() function with a different signature than the Appconfig model and it's complaining about it.  It seems that our options are to refactor the signature to match the heredity or change the function name.  Here's the thing: because of my refactoring, this method now only ever called 5 times in code and each time it's specifying the second parameter.  All the other calls have been replaced with `config('OSPOS')-><variable_name>` which gets populated in the app\Events\load_config.php class.

![image](https://user-images.githubusercontent.com/17935339/157028602-e7b107f7-7932-4e5d-bdc6-02a3146dfc86.png)

Thoughts? It seems the best course of action would be to refactor the method name to something like get_with_default().

**Comment by objecttothis** [comment]
> @jekkos @daN4cat @objecttothis
> 
> Can I help some part. I can give some times for it. Let's make new version in CodeIgniter 4.1.3 with bootstrap 5 as soon as possible.

Thank you for offering to help. Right now the most help is needed with the CI4 conversion. Do you have PHP experience? Let me think through a few parts of the code that you could help with.

**Comment by jekkos** [comment]
> @jekkos @daN4cat Here's another design decision issue. In CI4 all models need to extend Model. In this case Model has a get() function with a different signature than the Appconfig model and it's complaining about it. It seems that our options are to refactor the signature to match the heredity or change the function name. Here's the thing: because of my refactoring, this method now only ever called 5 times in code and each time it's specifying the second parameter. All the other calls have been replaced with `config('OSPOS')-><variable_name>` which gets populated in the app\Events\load_config.php class.
> 
> ![image](https://user-images.githubusercontent.com/17935339/157028602-e7b107f7-7932-4e5d-bdc6-02a3146dfc86.png)
> 
> Thoughts? It seems the best course of action would be to refactor the method name to something like get_with_default(). BTW there are a few of these types of methods in the code that we need to figure out.

It might be best to rename our method then to avoid any future conflict. Maybe we should call it get_value() as I don't think method overloading works in PHP? 
So if I folllow correctly  `$this-config->` will then now become `$this->config('OSPOS')->get_value($key, $value)`?

**Comment by bhojkamal** [comment]
> 

Hi, I have worked on many projects on CodeIgniter 3 for website development in the past. I developed a pharmacy software that is on CI3, jquery ajax, bootstrap. I've already developed a website in CodeIgniter 4. However, now my new projects are in Laravel. I have well experience in PHP. So, I can give some effort to bring new version of CodeIgniter live.

**Comment by objecttothis** [comment]
> > 
> 
> @objecttothis Hi, I have worked on many projects on CodeIgniter 3 for website development in the past. I developed a pharmacy software that is on CI3, jquery ajax, bootstrap. I've already developed a website in CodeIgniter 4. However, now my new projects are in Laravel. I have well experience in PHP. So, I can give some effort to bring new version of CodeIgniter live.

Excellent.  I'm glad for the help.  @jekkos I'd like to give @bhojkamal permissions to be able to check out the same branch that I am working on rather than need to worry about cloning it and merging his changes.  I think we need to add him as a contributor in order to give him permissions to check out and make commits to the branch.  Are you OK with this?

@bhojkamal There is already a Gitter dev group that we can use to collaborate on who will work on what parts. I just added you to the devroom.  Let's pick up the conversation there.  This way if other devs are wanting to jump in, they can not only see the conversation, but we can coordinate better.

**Comment by jekkos** [comment]
Yes sure go ahead on giving him at least write access to branches. All help is welcomed!

**Comment by objecttothis** [comment]
> It might be best to rename our method then to avoid any future conflict. Maybe we should call it get_value() as I don't think method overloading works in PHP? So if I folllow correctly `$this-config->` will then now become `$this->config('OSPOS')->get_value($key, $value)`?

OK I'll rename it.  Generally references to `$this->config->` become `config('OSPOS')->[variable_name]` and we use that throughout the code to draw from the cached variables.  In the rare case where we need the "default" that `appconfig->get_value($key, $value = '')` provides then that gets referenced like any other model function without config('OSPOS').  

`appconfig->save()` has been refactored from `appconfig->save($key, $value)` to `appconfig->save([$key => $value])` to match the signature of it's parent method.  The only thing that I'm doing different in the override is updating the config('OSPOS') cached corresponding value in the array so that it doesn't become stale.

**Comment by objecttothis** [comment]
@jekkos @daN4cat here is a bug in the code that PHP is complaining about and I need to resolve it to move forward with the CI4 conversion. The problem is that the code has a path that doesn't return anything.  Seems like either switching the elseif to an else statement or switching the entire block to a switch statement and adding a default would be the logical options.  Alternately if we wanted to go the route of throwing an exception we could do that as well, but it means wrapping the calling code in try/catch blocks and handling the exception.  Thoughts?
![image](https://user-images.githubusercontent.com/17935339/158175891-f5c7f9e8-4aa6-40a0-af58-7dc562758767.png)

**Comment by objecttothis** [comment]
Something like maybe?
```
		switch ($id)
		{
			case self::GOODS_SUPPLIER:
				return lang('Suppliers.goods');
			default:
			case self::COST_SUPPLIER:
				return lang('Suppliers.cost');
		}
```

**Comment by objecttothis** [comment]
To clarify, this is in the supplier model.

**Comment by daN4cat** [comment]
It's better to make the error visible and say that is an unknown string.

Not sure how comes that there is no default though.

**Comment by objecttothis** [comment]
> It's better to make the error visible and say that is an unknown string.
> 
> Not sure how comes that there is no default though.

Here is an example of a place where get_category_name() is called in the Suppliers controller.
![image](https://user-images.githubusercontent.com/17935339/158179121-73725f50-9c42-427d-a9ce-808103af24e3.png)

it's unclear to me how we would pass the error data in the json and therefore how we should pass the data back from the get_category_name() function in the event of bad data.  I think what you're saying we need to do is something like:

```
		switch ($id)
		{
			case self::GOODS_SUPPLIER:
				return lang('Suppliers.goods');
			case self::COST_SUPPLIER:
				return lang('Suppliers.cost');
			default:
				return lang('Suppliers.invalid_category');
		}
```
where Suppliers.invalid_category is a new phrase indicating that the category is invalid. I think the reason we haven't experienced averse effects from this is that the Supplier category is programatically set and would require data corruption or manual changing of the data to cause an issue.  

The problem with the above solution is that the calling get_row() in the Controller is just sending back the string via JSON in the AJAX call/response loop.  The view would also need to be updated to handle incorrect data.  If there is an example of how we are doing this in other places, then I can probably just mimic it.  Otherwise I'll likely need to create an issue that blocks #842 for someone to tackle

**Comment by objecttothis** [comment]
> Indeed it's always good to catch these unexpected cases explicitly, this avoids surprises.

OK, I'll throw an exception and wrap the calling code in a try/catch block.  From the catch, I'm not sure the best way to convey the error to the user.  I think it needs to be sent to the error log, but should also be added in a toast notification to the user that it failed.  If there is another part of the code that we are doing this already, can someone point me in that direction?

**Comment by objecttothis** [comment]
CI4 no longer has a built in utility to dump the database.  I think we just need to rewrite the function to use mysqldump.  Any takers? @opensourcepos/org @bhojkamal
![image](https://user-images.githubusercontent.com/17935339/158777803-78b77594-59ce-4b51-aedb-755560d78fdd.png)

**Comment by objecttothis** [comment]
CI4 no longer uploads images the same way. I've started the conversion.  Anyone want to finish it? @opensourcepos/org @bhojkamal 
![image](https://user-images.githubusercontent.com/17935339/158799863-b626f8f3-686c-4a51-b734-7c3815fe0ca9.png)

**Comment by bhojkamal** [comment]
> CI4 no longer uploads images the same way. I've started the conversion. Anyone want to finish it? @opensourcepos/org @bhojkamal ![image](https://user-images.githubusercontent.com/17935339/158799863-b626f8f3-686c-4a51-b734-7c3815fe0ca9.png)

I will see next week. I've done the uploading image one CI4 website.

**Comment by objecttothis** [comment]
@bhojkamal I changed the title back because the codebase I pulled from CI was 4.1.3 when I first started not 4.1.9.  It needs to be frozen at 4.1.3 until the conversion is complete and regressions have been mitigated.  Later we can upgrade to the latest release.

**Comment by objecttothis** [comment]
Update:
I continue to plug away at the changes.  @bhojkamal has been helping when he has time between his other projects.  If I don't finish before 15 May, I will need to take a 3 month hiatus due to travel.  During that time others can continue to work on the project.  As always, any help is welcome and will speed up the process.
@opensourcepos/org @BudsieBuds

**Comment by bhojkamal** [comment]
Hello, guys any update for CI4  upgrade?

**Comment by objecttothis** [comment]
> Hello, guys any update for CI4 upgrade?

Please see my post above: 
> If I don't finish before 15 May, I will need to take a 3 month hiatus due to travel.

**Comment by jekkos** [comment]
Ok I might check out the branch sooner or later to see in what state it is, maybe we can try to deploy and have some testing done.

**Comment by objecttothis** [comment]
> Ok I might check out the branch sooner or later to see in what state it is, maybe we can try to deploy and have some testing done.

Sounds good.  Feel free.  I will pick it up again in September when I'm back.

**Comment by objecttothis** [comment]
I've created the CodeIgniter4 label to make it easier to see and search for CI4 related issues.
